File: CxDLOCT/ForwardModel_PointScatterers_HighNA.py
def ForwardModel_PointScatterers_HighNA(amp, z, x, k_vect, xi, alpha, z_fp, z_ref, max_batch_size=None):
    # Number of points
    n_points = z.shape[2]
    
    # If not input batchSize calculate contribution from all points at once
    if max_batch_size is None:
        max_batch_size = n_points
    
    # Batch size
    batch_size = min(max_batch_size, n_points)
    
    # Number of batches of points
    n_batches = np.ceil(n_points / batch_size).astype(int)
    
    # Initialize output
    fringes = np.zeros((k_vect.shape[0], 1), dtype=k_vect.dtype)
    
    # Iterate batches
    for j in range(n_batches):
        # Calculate the contribution from this batch of points
        this_batch = np.unique(np.arange(batch_size) + (j * batch_size), return_index=True)[1]
        this_batch = np.take(this_batch, np.where(this_batch < n_points))
        
        # In this case we use sqrt((2*k)^2 - xi^2) where k is a vector
        this_fringes = (1 / (8 * np.pi) / \
          ((alpha / k_vect) ** 2 + (1j * (z[:, :, this_batch] - z_ref) / k_vect)) * \
          np.sum(np.exp(-1j * xi * x[:, :, this_batch] + \
          (1j * (z[:, :, this_batch] - z_ref) * np.sqrt((2 * k_vect) ** 2 - xi ** 2)) + \
          (- (xi * alpha / k_vect / 2) ** 2)), 1))
        
        # sum the contribution of this batch of scatterers, considering its
        # individual amplitudes, and sum to the overall contribution
        fringes = fringes + np.sum(amp[:, :, this_batch] * this_fringes, axis=2)
    
    return fringes

File: CxDLOCT/ForwardModel_PointScatterers_HighNA.m
function fringes = ForwardModel_PointScatterers_HighNA(amp, z, x, kVect, xi,...
  alpha, zFP, zRef, maxBatchSize)

  % Number of points
  nPoints = size(z, 3);
  % If not input batchSize calculate contribution from all points at once
  if nargin < 9
    maxBatchSize = nPoints;
  end
  % Batch size
  batchSize = single(min(maxBatchSize, nPoints));
  if isa(kVect,'gpuArray')
    batchSize = gpuArray(batchSize);
  end
  % Number of batches of points
  nBatches = ceil(nPoints / batchSize);
  % Initialize output
  fringes = zeros(size(kVect, 1), 1, 'like', kVect);
  % Iterate batches
  for j = 1:nBatches
    % Calculate the contribution from this batch of points
    thisBatch = min((1:batchSize) + (j - 1) * batchSize, nPoints);
    thisBatch = unique(thisBatch,'first');
    % In this case we use sqrt((2*k)^2 - xi^2) where k is a vector
    thisfringes = (1 / (8 * pi) ./ ...
      ((alpha ./ kVect) .^ 2 + (1i * (z(:, :, thisBatch) - zRef) ./ kVect)) .* ...
      sum(exp(-1i * xi .* x(:, :, thisBatch) + ...
      (1i * (z(:, :, thisBatch) - zRef) .* sqrt((2 * kVect) .^ 2 - xi .^ 2)) + ...
      (- (xi * alpha ./ kVect / 2) .^ 2)), 2));
    % sum the contribution of thi batch of scatteres, considering its
    % individual amplitudes, and sum to the overall contribution
    fringes = fringes + sum(amp(:, :, thisBatch) .* thisfringes, 3);
  end
end